## Gather

In [835]:
import pandas as pd
import numpy as np 

In [836]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [837]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [838]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [839]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [840]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [841]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [842]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [843]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [844]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [845]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [846]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [847]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [848]:
patients.sample(5)

patient_id assigned_sex given_name     surname                address  \
247         248         male     Tuukka  Leppäluoto     1886 Bicetown Road   
92           93         male    Michael       Smith    666 Whiteman Street   
424         425       female  Krystiana   Kozłowska      2935 Diamond Cove   
328         329       female       Anja      Hueber   3216 Lodgeville Road   
421         422       female     Armina       Sauvé  4977 Arlington Avenue   

            city state  zip_code        country  \
247     New York    NY   10011.0  United States   
92   Mount Holly    NJ    8060.0  United States   
424   Providence    RI    2906.0  United States   
328  Minneapolis    MN   55402.0  United States   
421  Little Rock    AR   72210.0  United States   

                                          contact  birthdate  weight  height  \
247      917-408-8855TuukkaLeppaluoto@teleworm.us   3/7/1978   211.0      73   
92           609-914-8473MichaelSmith@armyspy.com  9/17/1977   160.6      67   
424  KrystianaKozlowska@cuvox.de+1 (401) 419-2975  11/8/1921   123.6      63   
328       AnjaHueber@teleworm.us+1 (612) 342-6065  4/16/1987   151.8      65   
421            870-270-5502ArminaSauve@dayrep.com  10/8/1928   165.0      61   

      bmi  
247  27.8  
92   25.2  
424  21.9  
328  25.3  
421  31.2

In [849]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Dratchev       2
Ogochukwu      2
Johnson        2
Nilsen         2
Schiavone      2
Silva          2
Aranda         2
Parker         2
Woźniak        2
Liễu           2
Lương          2
Cabrera        2
Tạ             2
Kadyrov        2
Souza          2
Lund           2
Cindrić        2
Correia        2
Kowalczyk      2
Lâm            2
Gersten        2
Hueber         2
Collins        2
Bùi            2
Grímsdóttir    2
Batukayev      2
Berg           2
              ..
Chin           1
Györfy         1
Ščančar        1
Tikhonov       1
Cotton         1
Mata           1
Terrazas       1
Yudina         1
Radislav       1
Mironov        1
Lê             1
Írisarson      1
Holden         1
Yonatan        1
Jurković       1
Clapp          1
Karlsen        1
Onio           1
Amari          1
Lucchese       1
Luoma          1
Glockner       1
Gunnarsson     1
Markus         1
Baum           1
Koldenhof      1
Vesecká        1
Stanić        

In [850]:
patients.address.value_counts()

123 Main Street               6
648 Old Dear Lane             2
2476 Fulton Street            2
2778 North Avenue             2
192 Patton Lane               1
932 Memory Lane               1
2661 Russell Street           1
2785 Norma Lane               1
4380 Riverside Drive          1
3548 Hilltop Drive            1
3210 Hickory Lane             1
163 Hide A Way Road           1
3893 Eva Pearl Street         1
1886 Bicetown Road            1
4970 Heather Sees Way         1
3434 Holt Street              1
1984 Counts Lane              1
383 Marion Street             1
4789 Devils Hill Road         1
1067 Ray Court                1
494 Clarksburg Park Road      1
2074 Parrish Avenue           1
3682 Stiles Street            1
1736 Parrill Court            1
1424 Primrose Lane            1
2011 Eastland Avenue          1
4982 Wood Street              1
1270 Haul Road                1
1463 Martha Ellen Drive       1
4230 Ridenour Street          1
                             ..
576 Brow

In [851]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [852]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [853]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [854]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [855]:
sum(treatments.auralin.isnull())

0

In [856]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [857]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define

The missing `treatments` records are stored in a file named `treatments_cut.csv`.  
Read the `treatments_cut.csv` file as Dataframe, and concatenate it with `treatments_clean` to get the missing records. 

##### Code

In [858]:
# read the treatments_cut.csv 
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [859]:
# Join the missing data from the treatments_cut to treatments_clean 
treatments_clean = pd.concat([treatments_clean, treatments_cut])

##### Test

In [860]:
# Check the row counts 
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


In [861]:
# Check if there is any duplicates
treatments_clean[treatments_clean.duplicated(keep=False)]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
10      joseph     day  29u - 36u       -          7.7       7.19   
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
10            NaN  
136           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define

Recalculate the *hba1c_change* column: *hba1c_start* minus *hba1c_end*.

##### Code

In [862]:
# Caculate the hba1c change with hba1c start and hba1c end values. 
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start']-treatments_clean['hba1c_end']

##### Test

In [863]:
# Check the results
treatments_clean.sample(10)

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
251      aksel  vestergaard          -  42u - 38u         9.62       9.29   
84      furuta        osman  30u - 41u          -         7.52       7.18   
57      joseph       tucker  48u - 56u          -         7.67       7.30   
176      andré       nordin  40u - 47u          -         7.60       7.27   
98      idalia        moore  25u - 31u          -         8.34       7.90   
78       julia       werner  35u - 40u          -         7.58       7.24   
269     hiromu     horikawa          -  47u - 46u         7.77       7.28   
53       genet     fesahaye  25u - 35u          -         7.73       7.36   
67       bernd    schneider  48u - 56u          -         7.74       7.44   
41      mahmud      kadyrov          -  44u - 43u         7.53       7.11   

     hba1c_change  
251          0.33  
84           0.34  
57           0.37  
176          0.33  
98           0.44  
78           0.34  
269          0.49  
53           0.37  
67           0.30  
41           0.42

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

##### Code

In [864]:
# Extract phone number and email each 
#patients_clean['phone']= patients_clean['contact'].str.extract(r'(\d{3}[-\.\s]??\d{3}[-\.\s]??\d{4}|\(\d{3}\)\s*\d{3}[-\.\s]??\d{4}|\d{3}[-\.\s]??\d{4})', expand=False)
patients_clean['phone']= patients_clean.contact.str.extract(r'((?:\+*\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
#patients_clean['email']= patients_clean['contact'].str.extract(r"([a-zA-Z_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)",expand=False)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

In [865]:
# Drop the 'contact' column *Note: axis=1 denotes that we are referring to a column, not a row


##### Test

In [866]:
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone',
 'email']

In [867]:
# Check phone number
patients_clean.phone.sample(25)

299         208-826-1678
398    +1 (706) 777-4981
122         231-607-3625
320         262-565-9052
457         817-777-5289
37     +1 (605) 204-6572
72          504-289-1386
496       1 209 762 2320
34          207-825-8634
166         903-749-8046
132         570-698-4203
247         917-408-8855
206         806-323-2210
196         512-738-2609
50          570-819-1140
484         336-697-2005
441         559-693-6779
310       1 913 322 9114
412    +1 (513) 297-6359
180         318-417-5739
56     +1 (706) 497-0891
64     +1 (315) 362-5858
113       1 713 779 6516
392         630-252-5095
266       1 606 632 4327
Name: phone, dtype: object

In [868]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define

Unpivot the `treatments_clean` dataframe. 


##### Code

In [869]:
# unpivot the 'auralin' and 'novodra' columns
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'], 
                           value_vars=['auralin', 'novodra'], var_name='treatment', value_name='dose')

In [870]:
# split the 'dose' column into 'start_dose' and 'end_dose'
treatments_clean[['start_dose', 'end_dose']]=treatments_clean['dose'].str.split(pat="-", expand=True)

In [871]:
# extract only numbers from 'dose'
treatments_clean['start_dose'] = treatments_clean['dose'].str.extract(r'(\d*)')
treatments_clean['end_dose'] = treatments_clean['dose'].str.extract(r"[-](\s\d*)")

# or same result with below code 
#treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str

In [872]:
# drop the rows with '-' value in 'dose' column
treatments_clean.drop(treatments_clean[treatments_clean['dose']=='-'].index, inplace=True)

# or same result with below code
#treatments_clean = treatments_clean[treatments_clean.dose != "-"]

In [873]:
# drop the 'dose' column
treatments_clean.drop('dose', axis=1, inplace=True)

In [874]:
# reset index
treatments_clean.reset_index(drop=True, inplace=True)

##### Test

In [875]:
# check 
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  start_dose end_dose  
0         41       48  
1         33       36  
2         37       42  
3         31       38  
4         30       36

#### Adverse reaction should be part of the `treatments` table

##### Define

Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

##### Code

In [876]:
# merge 'adverse reactions' dataframe with 'treatments' dataframe
treatments_clean = pd.merge(adverse_reactions, treatments_clean, how='right', on=['given_name', 'surname'])

##### Test

In [877]:
# Check
treatments_clean

given_name        surname           adverse_reaction  hba1c_start  \
0          berta     napolitani  injection site discomfort         7.68   
1           lena           baer               hypoglycemia         7.70   
2         joseph            day               hypoglycemia         7.70   
3         joseph            day               hypoglycemia         7.70   
4         flavia     fiorentino                      cough         7.95   
5        manouck        wubbels          throat irritation         7.66   
6        jasmine          sykes               hypoglycemia         7.56   
7         louise        johnson               hypoglycemia         7.63   
8        albinca        komavec               hypoglycemia         7.89   
9            noe         aranda               hypoglycemia         7.51   
10         sofia      hermansen  injection site discomfort         8.90   
11         tegan        johnson                   headache         7.79   
12          abel        yonatan                      cough         7.88   
13     abdul-nur            isa               hypoglycemia         7.98   
14          leon         scholz  injection site discomfort         7.72   
15      gabriele        saenger               hypoglycemia         8.35   
16        jia li           teng                     nausea         7.66   
17         jakob       jakobsen               hypoglycemia         7.96   
18   christopher       woodward                     nausea         7.51   
19           ole       petersen               hypoglycemia         7.95   
20        finley       chandler                   headache         7.65   
21      anenechi          chidi               hypoglycemia         7.64   
22      miłosław     wiśniewski  injection site discomfort         7.51   
23         lixue          hsueh  injection site discomfort         9.21   
24         merci         leroux               hypoglycemia         8.98   
25          kang            mai  injection site discomfort         7.78   
26        elliot     richardson               hypoglycemia         7.56   
27       clinton         miller          throat irritation         7.79   
28        idalia          moore               hypoglycemia         8.34   
29        xiuxiu          chang               hypoglycemia         7.64   
..           ...            ...                        ...          ...   
320       yasmin         araujo                        NaN         7.82   
321       hiromu       horikawa                        NaN         7.77   
322          leo         vieira                        NaN         7.74   
323         john    teichelmann                        NaN         7.90   
324          thể          lương                        NaN         7.64   
325         meya        sjöberg                        NaN         7.77   
326         jane        citizen                        NaN         7.98   
327       angela     lavrentyev                        NaN         7.61   
328       edelma    villalpando                        NaN         7.99   
329       annika          vaara                        NaN         7.73   
330        chiho           higa                        NaN         7.71   
331    beatrycze        woźniak                        NaN         7.54   
332      firenze          fodor                        NaN         7.89   
333          zoe        wellish                        NaN         7.71   
334          una  traustadóttir                        NaN         8.00   
335        luboš          pecha                        NaN         7.79   
336        meaza         brhane                        NaN         7.70   
337        adlan       shishani                        NaN         7.84   
338        guðni      heimisson                        NaN         7.64   
339      eufemio        rosario                        NaN         7.54   
340     dalmacia         madrid                        NaN         7.67   
341        daimy          tromp

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define

Adverse reactions table is no longer needed so ignore that part.   
Convert the names in *given_name* and *surname* columns to titlecase in `treatments` table.   
Merge `patients` table and `treatments` table on *given_name* and *surname* columns to add *patients_id* column from `patients` table to `treatments` table. Drop *given_name* and *surname* columns from `treatments` table. 

##### Code

In [878]:
# Convert the names in 'given_name' and 'surname' columns to titlecase in 'treatments' table.
treatments_clean['given_name']=treatments_clean['given_name'].str.title()
treatments_clean['surname']=treatments_clean['surname'].str.title()

# Or you can convert these names to lower case in 'patients' table 
#patients_clean.given_name = patients_clean.given_name.str.lower()
#patients_clean.surname = patients_clean.surname.str.lower()

In [879]:
# Add 'patient_id' column to 'treatments' table 
# Note! we only need 'patient_id' column from 'patients_clean', so when merging, indicate the only columns we need. 
treatments_clean = pd.merge(treatments_clean, patients_clean[['patient_id', 'given_name', 'surname']], how='left', on=['given_name', 'surname'])

In [880]:
# Drop 'given_name' and 'surname' columns in treatments_clean 
treatments_clean.drop(['given_name', 'surname'], axis=1, inplace=True)

##### Test

In [881]:
# Check
treatments_clean

adverse_reaction  hba1c_start  hba1c_end  hba1c_change  \
0    injection site discomfort         7.68       7.21          0.47   
1                 hypoglycemia         7.70       7.40          0.30   
2                 hypoglycemia         7.70       7.19          0.51   
3                 hypoglycemia         7.70       7.19          0.51   
4                        cough         7.95       7.60          0.35   
5            throat irritation         7.66       7.40          0.26   
6                 hypoglycemia         7.56       7.18          0.38   
7                 hypoglycemia         7.63       7.32          0.31   
8                 hypoglycemia         7.89       7.46          0.43   
9                 hypoglycemia         7.51       7.17          0.34   
10   injection site discomfort         8.90       8.57          0.33   
11                    headache         7.79       7.43          0.36   
12                       cough         7.88       7.50          0.38   
13                hypoglycemia         7.98       7.53          0.45   
14   injection site discomfort         7.72       7.29          0.43   
15                hypoglycemia         8.35       7.97          0.38   
16                      nausea         7.66       7.32          0.34   
17                hypoglycemia         7.96       7.51          0.45   
18                      nausea         7.51       7.06          0.45   
19                hypoglycemia         7.95       7.55          0.40   
20                    headache         7.65       7.26          0.39   
21                hypoglycemia         7.64       7.31          0.33   
22   injection site discomfort         7.51       7.08          0.43   
23   injection site discomfort         9.21       8.80          0.41   
24                hypoglycemia         8.98       8.64          0.34   
25   injection site discomfort         7.78       7.45          0.33   
26                hypoglycemia         7.56       7.09          0.47   
27           throat irritation         7.79       7.40          0.39   
28                hypoglycemia         8.34       7.90          0.44   
29                hypoglycemia         7.64       7.23          0.41   
..                         ...          ...        ...           ...   
320                        NaN         7.82       7.36          0.46   
321                        NaN         7.77       7.28          0.49   
322                        NaN         7.74       7.36          0.38   
323                        NaN         7.90       7.58          0.32   
324                        NaN         7.64       7.22          0.42   
325                        NaN         7.77       7.42          0.35   
326                        NaN         7.98       7.60          0.38   
327                        NaN         7.61       7.14          0.47   
328                        NaN         7.99       7.56          0.43   
329                        NaN         7.73       7.34          0.39   
330                        NaN         7.71       7.30          0.41   
331                        NaN         7.54       7.17          0.37   
332                        NaN         7.89       7.55          0.34   
333                        NaN         7.71       7.30          0.41   
334                        NaN         8.00       7.50          0.50   
335                        NaN         7.79       7.45          0.34   
336                        NaN         7.70       7.36          0.34   
337                        NaN         7.84       7.37          0.47   
338                        NaN         7.64       7.24          0.40   
339                        NaN         7.54       7.26          0.28   
340                        NaN         7.67       7.21          0.46   
341                        NaN         9.41       8.94          0.47   
342                        NaN         7.68       7.36          0.32   
343                        NaN         7.78       7.39

In [882]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

23    patient_id
dtype: object

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert data type of *zip_code* from float to string using `astype()` method, and fill zeros at the beginning using `zfill()` method. 

##### Code

In [883]:
# Convert the datatype of "zip_code" column and pad '0' character
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.zfill(5)

# Same results with pad() method
# patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(width=5,fillchar='0')

In [884]:
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [885]:
# Check the data type of 'zip_code' 
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null object
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone           491 non-null object
email           491 non-null object
dtypes: float64(2), int64(2), object(12)
memory usage: 63.0+ KB


In [886]:
patients_clean.zip_code.sample(10)

127    98541
141    75207
181    81637
12     24504
316    01002
456    23602
414    45380
255    54739
97     68324
52     50322
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace 27 to 72 for the height of Neudorft using `replace()` method.

##### Code

In [887]:
# Replace 27 to 72
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [888]:
# Check if there is still 27 
patients_clean[patients_clean.height==27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, contact, birthdate, weight, height, bmi, phone, email]
Index: []

#### Full state names sometimes, abbreviations other times

##### Define


##### Code

In [889]:
# Check value counts for "state" column which have full name
patients_clean[patients_clean.state.str.len() != 2].state.value_counts()

California    36
New York      25
Illinois      14
Florida       13
Nebraska       4
Name: state, dtype: int64

In [890]:
# Mappiing from full state name to abbreviations in those 5 values
state_abb = {'California' : 'CA', 
             'New York' : 'NY', 
             'Illinois': 'IL',
             'Florida': 'FL',
             'Nebraska' : 'NE'
            }

# Make a function to apply 
def full_to_abb(patient):
    if patient['state'] in state_abb.keys(): 
        abb = state_abb[patient['state']]
        return abb
    else:
        return patient['state']

# Apply the function 
patients_clean['state'] = patients_clean.apply(full_to_abb, axis=1)

##### Test

In [891]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
MI    13
LA    13
OK    13
NJ    12
VA    11
MS    10
WI    10
MN     9
AL     9
IN     9
TN     9
KY     8
WA     8
NC     8
MO     7
KS     6
ID     6
NE     6
NV     6
CT     5
SC     5
IA     5
ND     4
ME     4
AR     4
RI     4
AZ     4
CO     4
DE     3
OR     3
SD     3
WV     3
MD     3
MT     2
DC     2
VT     2
AK     1
NH     1
WY     1
NM     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'

##### Code

In [892]:
# Replace "Dsvid" with "David"
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [893]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country  \
8  Kansas City    MO    64105  United States   

                                   contact birthdate  weight  height   bmi  \
8  816-265-9578DavidGustafsson@armyspy.com  3/6/1937   163.9      66  26.5   

          phone                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define

Convert Datatypes
- Already converted : *zip_code*
- From 'object' to 'category' : *assigned_sex*, *state*
- From 'object' to 'datetime' : *birthdate*
- From 'object' to 'int' : *start_dose*, *end_dose* (Already strip the letter 'u')


##### Code

In [894]:
# Check the data types of "patients_clean"
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null object
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone           491 non-null object
email           491 non-null object
dtypes: float64(2), int64(2), object(12)
memory usage: 63.0+ KB


In [895]:
# Check the data types of "treatments_clean"
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 8 columns):
adverse_reaction    35 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start_dose          350 non-null object
end_dose            350 non-null object
patient_id          344 non-null float64
dtypes: float64(4), object(4)
memory usage: 24.6+ KB


In [896]:
# Object to Category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

In [897]:
# Object to Datetime 
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

In [898]:
# Object to Int
treatments_clean.start_dose = treatments_clean.start_dose.astype('int64')
treatments_clean.end_dose = treatments_clean.end_dose.astype('int64')

##### Test

In [899]:
# Check datatypes
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 16 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone           491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(9)
memory usage: 57.8+ KB


In [900]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 8 columns):
adverse_reaction    35 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
start_dose          350 non-null int64
end_dose            350 non-null int64
patient_id          344 non-null float64
dtypes: float64(4), int64(2), object(2)
memory usage: 24.6+ KB


In [901]:
patients_clean.sample(3)

patient_id assigned_sex given_name      surname               address  \
311         312       female      Erica    MacDonald  2011 Eastland Avenue   
85           86         male    Burrell      Mathieu     3816 Grand Avenue   
72           73         male      Søren  Frederiksen      4852 Rose Avenue   

             city state zip_code        country  \
311  Philadelphia    MS    39350  United States   
85      Kissimmee    FL    34746  United States   
72       Metairie    LA    70001  United States   

                                     contact  birthdate  weight  height   bmi  \
311   EricaMacDonald@teleworm.us601-389-7682 1937-05-04   171.4      63  30.4   
85    407-662-9389BurrellMathieu@teleworm.us 1936-05-16   203.5      71  28.4   
72   SrenFrederiksen@armyspy.com504-289-1386 1931-02-12   214.5      65  35.7   

            phone                        email  
311  601-389-7682   EricaMacDonald@teleworm.us  
85   407-662-9389   BurrellMathieu@teleworm.us  
72   504-289-1386  SrenFrederiksen@armyspy.com

In [902]:
treatments_clean.sample(3)

adverse_reaction  hba1c_start  hba1c_end  hba1c_change treatment  \
94               NaN         7.84       7.57          0.27   auralin   
292              NaN         7.98       7.51          0.47   novodra   
3       hypoglycemia         7.70       7.19          0.51   auralin   

     start_dose  end_dose  patient_id  
94           24        32       121.0  
292          30        32       266.0  
3            29        36        70.0

#### Multiple phone number formats

##### Define

Strip all non-digit characters including "+", "-", " ", "(", ")".  
Pad the phone number with a '1' if the length of the number is 10 digits (add country code).

##### Code

In [907]:
# Replace all non-digit characters with ""
patients_clean.phone.replace(r'[^\d]',"",regex=True, inplace=True)

In [910]:
# Pad the phone number with a '1'
patients_clean.phone = patients_clean.phone.str.pad(width=11,fillchar='1')

##### Test

In [911]:
# Check the length of 'phone' column
patients_clean.phone.str.len().value_counts()

11.0    491
Name: phone, dtype: int64

In [912]:
patients_clean.phone.sample(10)

422    16414759654
53     16173175055
342    18606184560
376    15016364058
396    18633863795
277    11234567890
263    12127824151
118    13376854885
269            NaN
129    16313707406
Name: phone, dtype: object

#### Default John Doe data

##### Define
As there are several duplicated rows for "John Doe" having same name and adress and other info, remove the data.

##### Code

In [915]:
# Remove the rows with surname of "Doe"
patients_clean = patients_clean[patients_clean.surname!="Doe"]

##### Test

In [919]:
# There should be no "Doe" records
patients_clean[patients_clean.surname=="Doe"]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, contact, birthdate, weight, height, bmi, phone, email]
Index: []

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [935]:
# Remove rows with nicknames
patients_clean = patients_clean[~(patients_clean.address.duplicated() & patients_clean.address.notnull())] # ~(tilde) means not

##### Test

In [937]:
# Your testing code here
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  \
24   Port Jervis    NY    12771  United States   
432      Houston    TX    77020  United States   

                                        contact  birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707 1985-08-01   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438 1962-11-25   185.2      67   

      bmi        phone                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [938]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country                               contact  \
97    NE    68324  United States  PatrickGersten@rhyta.com402-848-4923   

    birthdate  weight  height   bmi        phone                     email  
97 1954-05-03   138.2      71  19.3  14028484923  PatrickGersten@rhyta.com

In [939]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country                                 contact  \
131    WV    25962  United States    304-438-2648SandraCTaylor@dayrep.com   
426    FL    33179  United States  305-434-6299RogelioJTaylor@teleworm.us   

     birthdate  weight  height   bmi        phone                       email  
131 1960-10-23   206.1      64  35.4  13044382648    SandraCTaylor@dayrep.com  
426 1992-09-02   186.6      69  27.6  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [943]:
# Multiply 2.20462 to get the kgs from lbs
patients_clean.loc[patients_clean.surname=='Zaitseva', 'weight'] = patients_clean.weight * 2.20462

##### Test

In [946]:
# Test : 48.8 shouldn't be the lowest
patients_clean.weight.sort_values()

459    102.100000
335    102.700000
74     103.200000
317    106.000000
171    106.500000
51     107.100000
210    107.585456
270    108.100000
198    108.500000
48     109.100000
478    109.600000
141    110.200000
38     111.800000
438    112.000000
14     112.000000
235    112.200000
307    112.400000
191    112.600000
408    113.100000
49     113.300000
326    114.000000
338    114.100000
253    117.000000
321    118.400000
168    118.800000
1      118.800000
350    119.000000
207    119.200000
265    120.000000
341    120.300000
          ...    
332    224.000000
12     224.200000
252    224.200000
222    224.800000
166    225.300000
111    225.900000
101    226.200000
150    226.600000
88     227.700000
352    227.700000
428    227.700000
13     228.400000
339    229.000000
182    230.300000
121    230.800000
257    231.700000
395    231.900000
246    232.100000
219    237.800000
11     238.700000
50     238.900000
441    239.100000
499    239.600000
439    242.000000
487    242